### DISCUSSION TITLE
Does it worth to use f16 instead of f32 ?

### DISCUSSION THREAD

**Discussion by tlecoz** [gpuweb/gpuweb#4122] [Q&A] [answered]
I use f32 because I used to use f32 and everything works well , so I'm just curious.
What are the pros/cons of f16. 

I understand that it's twice lighter than f32 in the buffer but I'm not sure to understand the price to paid for this improvment. 

What are the limitation that come with f16 usage ? 
May it be crucial to use it in term of performance on mobile device or it's not a question of performance at all? 

I have no idea at all how much important f16 can be. 
Can you give me some enlightment please :)

**Comment by Kangz** [comment, marked as answer]
F16 has multiple advantages:

 - it uses less memory traffic when reading / storing from buffers / textures. Memory traffic is often the limiting factor in GPU applications, so that's a good thing to optimize for.
 - Some GPUs are able to do f16 arithmetic at a double rate of f32.
 - Some GPUs are able to use a single "32 bit register" for two F16s, meaning shaders take less register and can have more instances running in parallel.

This blog post might be of interest: https://therealmjp.github.io/posts/shader-fp16/

The drawback of F16 is that it isn't always available and it only an optional feature in WebGPU.

**Reply by tlecoz** [reply]
Thank you for your message ! 

How can I know if the computer that runs the shader support f16 ?

**Reply by ben-clayton** [reply]
You will need to find [`shader-f16`](https://www.w3.org/TR/webgpu/#shader-f16) in [`GPUDevice.features`](https://www.w3.org/TR/webgpu/#gpudevice) - however I do not believe any browser implements this right now without additional command line arguments.

Chromium based browsers will be exposing this feature once we have a working DXIL backend (in progress) and good CTS coverage (also in progress). I'm afraid I do not have any precise timelines for you right now, but we hope to have this enabled sometime later this year.

**Comment by tlecoz** [comment]
Ok thank you ! 

"however I do not believe any browser implements this right now without additional command line arguments."
There is no chrome://flag ?

**Comment by ben-clayton** [comment]
I don't think so. Even if it did, on Windows it's dependent on [DXC](https://github.com/microsoft/DirectXShaderCompiler) being found, which is not a standard part of the OS, nor currently included in the Chrome install - so it's unlikely to be easily enabled right now.

**Comment by tlecoz** [comment]
Ok, it doesn't matter for the moment, I just was curious about it but the mystery is solved now :) 
Thank you again !
